desktop_server/src/com/mygdx/game/server/ecs/system/VisibilitySystem.java

```diff
@@ -0,0 +1,81 @@
+package com.mygdx.game.server.ecs.system;
+
+import com.artemis.ComponentMapper;
+import com.artemis.EntitySubscription;
+import com.artemis.annotations.All;
+import com.artemis.annotations.AspectDescriptor;
+import com.artemis.systems.IteratingSystem;
+import com.badlogic.gdx.utils.Bits;
+import com.badlogic.gdx.utils.IntMap;
+import com.mygdx.game.core.ecs.component.Coordinates;
+import com.mygdx.game.core.ecs.component.SubField;
+import com.mygdx.game.server.ecs.component.ChangeSubscribers;
+import com.mygdx.game.server.ecs.component.SightlineSubscribers;
+import lombok.extern.java.Log;
+
+import javax.inject.Inject;
+
+@All({Coordinates.class, SightlineSubscribers.class})
+@Log
+public class VisibilitySystem extends IteratingSystem {
+
+  @AspectDescriptor(all = {Coordinates.class, ChangeSubscribers.class}, exclude = { SubField.class })
+  private EntitySubscription allThatCanBePerceived;
+  private final IntMap<Bits> entityToNewChangeSubscribers = new IntMap<>();
+  private ComponentMapper<SightlineSubscribers> sightlineSubscribersMapper;
+  private ComponentMapper<ChangeSubscribers> changeSubscribersMapper;
+  private ComponentMapper<Coordinates> coordinatesMapper;
+
+  @Inject
+  public VisibilitySystem() {
+    super();
+  }
+
+  private Bits obtainNewChangeSubscribers(int entityId) {
+    var subscribers = entityToNewChangeSubscribers.get(entityId);
+    if (subscribers == null) {
+      subscribers = new Bits();
+      entityToNewChangeSubscribers.put(entityId, subscribers);
+    }
+    return subscribers;
+  }
+
+  @Override
+  protected void begin() {
+    log.info("process VisibilitySystem ");
+    entityToNewChangeSubscribers.clear();
+  }
+
+  @Override
+  protected void process(int perceiver) {
+    var coordinates = coordinatesMapper.get(perceiver);
+    var sightlineRadius = sightlineSubscribersMapper.get(perceiver).getSightlineRadius();
+    var sightlineSubscribers = sightlineSubscribersMapper.get(perceiver).getClients();
+    log.info(coordinates.toString());
+    for (int i = 0; i < allThatCanBePerceived.getEntities().size(); i++) {
+      var perceivableCoords = coordinatesMapper.get(allThatCanBePerceived.getEntities().get(i));
+      var dst2 = Math.pow(coordinates.getX() - perceivableCoords.getX(), 2) + Math.pow(coordinates.getY() - perceivableCoords.getY(), 2);
```

Comment:
Jako że to hexagonalna mapa to liczenie odległości wygląda trochę inaczej, w bocie robiłem najpierw konwersje na 'cube coordinates' potem liczyłem odległość

var ax = coordinates.getX() - (coordinates.getY() - (coordinates.getY() % 2)) / 2;
var ay = coordinates.getY();
var az = -ax-ay;

var bx = perceivableCoords.getX() - (perceivableCoords.getY() - (perceivableCoords.getY() % 2)) / 2;
var by = perceivableCoords.getY();
var bz = -bx-by;

var distance =  Math.max(Math.max(Math.abs(ax - bx), Math.abs(ay - by)), Math.abs(az - bz));

Reply:
hura, to to wkleję jutro